Add ws_proxy option for connecting gateway through a proxy.

zzzzRuby opened the pull request:
#2479

GnomedDev committed:
Remove temporary `Vec` usage in the `http` module (#2624, #2646)

This avoids having to allocate to store fixed length (replaced with normal
array) or fixed capacity (replaced with `ArrayVec`) collections as vectors for
the purposes of putting them through the `Request` plumbing.

Slight behavioral change - before, setting `params` to `Some(vec![])`
would still append a question mark to the end of the url. Now, we check
if the params array `is_empty` instead of `is_some`, so the question
mark won't be appended if the params list is empty.

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

GnomedDev committed:
Remove deprecated items (#2645)

GnomedDev committed:
Remove more deprecated items (#2651)

GnomedDev committed:
Remove `*_arc` methods (#2654)

These are unnecessary. Accepting `impl Into<Arc<T>>` allows passing either `T` or `Arc<T>`.

GnomedDev committed:
Clean up `ShardManager`/`ShardQueuer`/`ShardRunner` (#2653)

GnomedDev committed:
Put `Message::thread` behind a `Box` (#2658)

This trades a heap allocation for messages sent along with thread
creation for `Message`'s inline size dropping from 1176 bytes to 760
bytes,

GnomedDev committed:
Replace `Vec` and `String` with `FixedArray` and `FixedString` for all models (#2656)

This shrinks type sizes by a lot; however, it makes the user experience slightly
different:

- `FixedString` must be converted to String with `.into()` or `.into_string()`
  before it can be pushed to, but dereferences to `&str` as is.
- `FixedArray` must be converted to `Vec` with `.into()` or `.into_vec()`
  before it can be pushed to, but dereferences to `&[T]` as is.

The crate of these types is currently a Git dependency, but this is fine for
the `next` branch. It needs some basic testing, which Serenity is perfect for,
before a release will be made to crates.io.

zzzzRuby committed:
Add ws_proxy option for connecting gateway through a proxy.

GnomedDev committed:
Fix truncation when using `FixedString<u8>` where multibyte characters push over the limit (#2660)

Kerollmops committed:
Encode timestamps in RFC3339 (#2665)

Fixes #2664

ARandomDev99 committed:
Switch to `i64` in methods associated with `CreateCommandOption` (#2668)

This commit:

- switches from `u64` to `i64` in `CreateCommandOption::min_int_value` and
`CreateCommandOption::max_int_value` to accommodate negative integers in
Discord's integer range (between -2^53 and 2^53). Values outside this
range will cause Discord's API to return an error.
- switches from `i32` to `i64` in `CreateCommandOption::add_int_choice` and
`CreateCommandOption::add_int_choice_localized` to accommodate Discord's
complete integer range (between -2^53 and 2^53). Values outside this
range will cause Discord's API to return an error.

GnomedDev committed:
Get rid of the duplicate users cache (#2662)

This cache was just duplicating information already present in `Guild::members`
and therefore should be removed.

This saves around 700 MBs for my bot (pre-`FixedString`).

This has to refactor `utils::content_safe` to always take a `Guild` instead
of`Cache`, but in practice it was mostly pulling from the guild cache anyway
and this means it is more likely to respect nicknames and other information,
while losing the ability to clean mentions from DMs, which do not matter.

GnomedDev committed:
Change `Embed::fields` to use `FixedArray` (#2674)

`Embed::fields` previously had to stay as a `Vec` due to `CreateEmbed` wrapping
around it, but by implementing `Serialize` manually we can overwrite the
`Embed::fields` with a normal `Vec`, for a small performance hit on
serialization while saving some space for all stored `Embed`s.

GnomedDev committed:
Use `FixedArray` and `FixedString` in model enums (#2675)

Simply missed these when finding and replacing.

GnomedDev committed:
Bitpack boolean fields using fancy proc macro (#2673)

This uses the `bool_to_bitflags` macro to remove boolean (and optional boolean)
fields from structs and pack them into a bitflags invocation, so a struct with
many bools will only use one or two bytes, instead of a byte per bool as is.

This requires using getters and setters for the boolean fields, which changes
user experience and is hard to document, which is a significant downside, but
is such a nice change and will just become more and more efficient as time goes
on.

GnomedDev committed:
Use `nonmax::NonMax*` types in `Option`s wherever possible (#2681)

This swaps fields that store `Option<Int>` for `Option<NonMaxInt>` where the
maximum value would be ludicrous. Since `nonmax` uses `NonZero` internally,
this gives us niche optimisations, so model sizes can drop some more.

I have had to include a workaround for [#17] in `optional_string` by making my
own `TryFrom<u64>`, so that should be removable once that issue is fixed.

[#17]: https://github.com/LPGhatguy/nonmax/issues/17

GnomedDev committed:
Remove unused warning `#[allow(...)]`s (#2682)

A couple of clippy bugs have been fixed and I have shrunk model
sizes enough to make `clippy::large_enum_variant` go away.

GnomedDev committed:
Get rid of `unsafe` (#2686)

A discord bot library should not be using the tools reserved for low
level OS interaction/data structure libraries.

GnomedDev committed:
Swap Git dependencies for crates.io published versions (#2685)

GnomedDev committed:
Swap Id from `NonZero` to `NonMax` (#2689)

Discord seems to internally default Ids to 0, which is a bug whenever
exposed, but this makes ID parsing more resilient. I also took the
liberty to remove the `From<NonZero*>` implementations, to prevent future
headaches, as it was impossible to not break public API as we exposed
`NonZero` in `*Id::parse`.

GnomedDev committed:
Rewrite builders to take `Cow`s (#2688)

GnomedDev committed:
Store `Request::params` as a slice instead of owning `ArrayVec` (#2694)

This,
1. shrinks the size of Request, when copied around, as it doesn't have
to store the max capacity at all times
2. shrinks llvm-lines (compile time metric) for my bot in debug from
`1,153,519` to `1,131,480` as no monomorphisation has to be performed
for `MAX_PARAMS`.

mkrasnitski committed:
Revert `Request::params` back to `Option` (#2695)

Follow-up to #2694.

When `Request::params` was made into an ArrayVec, the `Option` around it
was removed in order to avoid having to add a turbofish on `None` to
specify the value of `MAX_PARAMS`. Also, `Request::new` also needed to
be changed so that the value of `MAX_PARAMS` could be inferred. Now that
the field is a slice again, we can wrap it in `Option` again (at no cost
to size, thanks to niche opts).

We ensure we never store the redundant `Some(&[])` by checking for an
empty slice and storing `None` instead. This way, we ensure we never
read an empty slice out of the `Some` variant.

GnomedDev committed:
Remove unnecessary `Vec` in `WsClient` (#2705)

GnomedDev committed:
Feature gate `#[instrument]` macros  (#2707)

The instrument macros generate 2% of Serenity's release mode llvm-lines,
and are proc-macros so hurt compile time in that way, so this limits
them to opt-in. This commit also fixes the issues that the instrument macro
was hiding, such as results that didn't ever error and missing
documentation.

GnomedDev committed:
Remove `impl Into<Option>` (#2701)

This signature is hard to use as `None` cannot infer the type of the
generic. I also replaced `Option<u8>` with `Option<NonMaxU8>` as it's
more efficient and will make the user think of the maximum value.

GnomedDev committed:
Get rid of `IntoIterator` generics wherever they are inefficent (#2698)

This removes inefficient `IntoIterator` generics and instead takes what is
actually required. I also reworked `reorder_channels` to allow for keeping the
generic, as it actually does only just need iterator semantics.

GnomedDev committed:
Clean up `Ratelimiter` (#2687)

Previously, someone assumed that `Ratelimiter` was going to be cloned, so
put a ton of `Arc`s everywhere. This was unneeded, and before dashmap,
so the buckets were also stored massively inefficiently. This fixes all
that.

I had to shuffle around the `Ratelimit` methods a little bit to return
their sleep time instead of sleeping themselves, so I didn't have to
hold a dashmap lock over an `.await`.

GnomedDev committed:
Handle overflow checks in central locations (#2697)

This removes multiple error variants and overall cleans up the codebase
by moving overflow checks into two `ModelError` variants.

GnomedDev committed:
Shrink size and clean up `Error` (#2700)

Shrinks `size_of::<Error>` from 136 bytes to 64 bytes, while removing unused
variants. This will improve performance for any method returning
`Result<T>` where `T` is less than the size of `Error` as both `Result`'s
`Ok` and `Err` have to be allocated stack space.

GnomedDev committed:
Remove manual `#[inline]` attributes (#2702)

The compiler knows best as inlining is quite complicated. This should
help with compile times, significantly.

GnomedDev committed:
Remove `Into<*Id>` and `AsRef<str>` (#2704)

`Into<*Id>` is often entirely unnecessary, and at worst hides the type of
the Id with hard-coded Ids. This fixes this by just using the Id type
itself, which also helps with monomorphisation bloat.

I also removed `AsRef<str>`, as this can just be `&str` without
requiring a generic and without requiring a call to as_ref in the
function, while also not accidentally taking ownership of a `String`
passed in misleadingly.

mkrasnitski committed:
Remove `Event::Unknown` variant (#2708)

If deserialization fails, we should allow the specific deserialization error to
be logged so that it can be reported to the Serenity maintainers. As it stands,
any events that fail to deserialize normally will "succeed" when deserialized
into the `Unknown` variant, throwing away any extra information about the
error.

Because deserialization errors are immediately logged when they happen,
explicitly whitelist `Error::Json` in `Shard::handle_event` to prevent
duplicate logs.

cheesycod commented:
This would help me as well. I forked nirn to hackily replace the ``ws`` url.

GnomedDev commented:
I'm kinda wondering why this is required at all, can't the gateway url just be exposed (if not already) to point to the proxy, instead of having this proxy option, and why is the manual http-over-tcp required? None of this is commented or explained.

zzzzRuby commented:
> I'm kinda wondering why this is required at all, can't the gateway url just be exposed (if not already) to point to the proxy, instead of having this proxy option, and why is the manual http-over-tcp required? None of this is commented or explained.

Sorry I'm new to opensource.
In some network environments like Chinese networks, there must be a proxy to connect with discord gateway. There're two options to do this, the first one is open a tunnel to the proxy likes this pr does, the second one is creating a custom gateway and redirect all message through a tunnel to the proxy. The second one seems too complex.
As to the manual http-over-tcp, I'm not sure if I can add a new dependency in Cargo.toml.

GnomedDev commented:
Can you not perform this proxying system wide, instead of coding support for it into every application? I'm honestly not familar with this.

zzzzRuby commented:
> Can you not perform this proxying system wide, instead of coding support for it into every application? I'm honestly not familar with this.

Normally, libraries like libcurl or reqwest get a default proxy from environment or system. But tungstenite doesn't support this and they seem have no interest in this (snapview/tungstenite-rs#177), the only choice to perform a proxy system wide is proxychains-like-tools. But I don't think proxychains is a good idea, it depends on linking libc dynamically, so on some distro like alpine it's hard to use with rust program and sometimes not work.

GnomedDev commented on src/client/mod.rs:
```diff
@@ -293,6 +295,23 @@ impl ClientBuilder {
+    }
+
+    /// Remove websocket proxy.
+    pub fn no_ws_proxy(mut self) -> Self {
+        self.ws_proxy = None;
+        self
+    }
```
Can this method be removed, I don't see a useful point to it.

zzzzRuby replied:
Removed it.

GnomedDev commented on src/client/mod.rs:
```diff
@@ -351,6 +371,7 @@ impl IntoFuture for ClientBuilder {
                 #[cfg(feature = "voice")]
                 voice_manager: voice_manager.as_ref().map(Arc::clone),
                 ws_url: Arc::clone(&ws_url),
+                ws_proxy,
```
Can this be wrapped in `Arc` like `ws_url`, and preprocessed into `Url` to fail here if the user provides an invalid URL.

zzzzRuby replied:
Yes it can. Finished

mkrasnitski replied:
`ws_url` is only parsed into a `Url` inside of the `gateway::shard::connect` helper function. I think for consistency's sake the same should be done with `ws_proxy` - a followup PR could move the parsing logic for both to be earlier in the initialization stage.

GnomedDev commented on src/gateway/ws.rs:
```diff
@@ -101,13 +108,72 @@ const TIMEOUT: Duration = Duration::from_millis(500);
 
 impl WsClient {
-    pub(crate) async fn connect(url: Url) -> Result<Self> {
+    async fn connect_with_proxy_async(
```
This whole method should be replaced with something using `reqwest`, which is already in the library.

Itsusinn replied:
> This whole method should be replaced with something using `reqwest`, which is already in the library.

Reqwest doesn't expose API that can be used in websocket proxy. But there is an implementation that borrows from reqwest's design.
https://github.com/snapview/tungstenite-rs/issues/177#issuecomment-1352808621

zzzzRuby replied:
Just like [Itsusinn](https://github.com/Itsusinn) said. I replaced the http connect code with async-http-proxy[https://crates.io/crates/async-http-proxy] which is mentioned in that issue.

GnomedDev reviewed (commented):
Okay, I understand now, the situation sucks (this should be on tungstenite) but if we can make a positive impact lets go for this

GnomedDev committed:
Use `NonMax` to signal builder range constraints (#2703)

GnomedDev committed:
Stop using `Value` in error parsing (#2710)

`Value` is bad to use and is slow. I also used `Arc<str>` to store `path` as
this is cloned a couple times.

GnomedDev committed:
Fix new clippy lints (#2713)

GnomedDev committed:
Update to `small-fixed-array` v0.2 (#2711)

`small-fixed-array` v0.2 removes implicit truncation and makes all
truncations explicit.

I also changed FixedString initialization from literals to use
`from_str_trunc` to take advantage of SSO if possible.

GnomedDev committed:
Fix `@everyone` role deserialisation (#2716)

The `@everyone` role sometimes has a role position of -1, but not always!

GnomedDev committed:
Fix `GuildChannel` deserialisation with `user_limit` (#2715)

This is documented as max 99 for voice channels, but 10k for stage
channels.

Milo123459 committed:
Fix maximum description length to the correct size (#2718)

GnomedDev committed:
Remove dashboard example (#2714)

This example depended on rillrate, which has been archived for over a
year.

arqunis committed:
Fix compilation from rebase

GnomedDev committed:
Remove deprecated `absolute_ratelimits` feature (#2719)

mkrasnitski committed:
Implement `max_concurrency` support when starting shards (#2661)

When a session is first kicked off and we need to start many shards, we
can use the value of `max_concurrency` to kick off multiple shards in
parallel.

GnomedDev committed:
Remove typemap without making everything generic (#2720)

Replaces the typemap with `dyn Any`, which is what typemap used
internally anyway. This gives the user more control over locking,
removes boilerplate, and is less confusing for beginners to learn.

GnomedDev committed:
Provide `Arc<Data>` instead of `&Data` (#2722)

This incurs an extra ref count, but this should be negligible for the
more permissive type. An explanation of how this works is that:

- `Arc` is just a shared pointer to `ArcInner`
- `ArcInner` stores the concrete Data type
- Each `Arc` stores the dyn vtable ptr
- `Arc` checks the vtable ptr to get the type id, then it can just give
out a new `Arc` referring to the `ArcInner` concretely.

GnomedDev committed:
Add `Client::try_data` to fallibly fetch the data type (#2723)

This is mostly for usage in `poise` to provide a nicer error message in
its situation. I also changed `Client::data` to provide `Arc` as well.

GnomedDev committed:
Fix voice manager initialisation (#2727)

mkrasnitski committed:
Remove some pedantic lints from the whitelist (#2728)

This also removes the impls of `From<{i32,u64}> for Colour`, as it only
makes sense to convert infallibly from `u32`.

GnomedDev committed:
Reduce unknown events to debug level, again (#2730)

GnomedDev committed:
Remove patch left in for `current` to stay non-breaking (#2732)

arqunis committed:
Fix compilation from rebase

mkrasnitski committed:
Fix testing example

mkrasnitski committed:
Remove `PartialEq` implementations from all builders (#2734)

Because builder fields are all private, it doesn't make sense to compare
them directly. From a user's perspective, they have no idea what the
builder might contain. This also lets us change their internals easier,
without worrying about equality.

GnomedDev committed:
Remove various instances of `impl AsRef<T>` (#2736)

This should improve compile times even more, by compiling a lot more
methods in Serenity, instead of the user's code.

GnomedDev committed:
Remove `simd-json` support (#2735)

`simd-json` is not a performance improvement for Serenity.

It has a ton of sketchy uses of `unsafe`, increases the maintenance burden, and
leads to Serenity containing a json compatibility layer(!).

This commit removes all that, although a faster json decoding implementation is
on the table in the future if the compatibility layer is kept out of tree and
it is measurably faster for Serenity's usecase.

GnomedDev committed:
Fix massive code size of `client::dispatch` (#2739)

This replaces fancy iter stuff with just writing the code, and therefore
reduces the code size significantly and spawns less tasks.

GnomedDev committed:
Fix RoleTags deserialisation after removal of simd-json (#2742)

There was a comment here that said it was called by simd-json, so I
removed the function, but it turns out that serde_json also calls it.

GnomedDev committed:
Avoid cloning FullEvent in dispatch code (#2740)

Avoids cloning entire event structs when dispatching. To reduce code
size and allocations all the state is bundled up into a temporary struct
called DispatchContext, but this isn't user facing at all. This
regresses code size a bit, but is worth it for runtime performance, and
probably reduces actual code sizes as `FullEvent` doesn't have to be
moved around as much.

GnomedDev committed:
Remove the builder trait and only take required arguments (#2741)

Removes unnecessary trait, which only limited the signatures to always
take `CacheHttp` and required the async_trait boxing and compile time
cost. This also makes a lot less methods generic, improving compile
times.

GnomedDev committed:
Use Serializer::collect_seq in serialize impls (#2744)

GnomedDev committed:
Convert enum number into wrapper struct to save type sizes (#2746)

Saves a couple bytes in each model by replacing the enums with a simpler
and more accurate representation, a newtype with named statics.
https://www.diffchecker.com/ohzayL9Z/

GnomedDev committed:
Use Arc<str> to store the token (#2745)

`SecretString` is nice to have, but this is more efficient, and
`SecretString` is defense in depth we can do without.

GnomedDev committed:
Remove standard framework (#2731)

This PR carries out the removal after the deprecation made on current.

GnomedDev committed:
Re-add Secret to bot token (#2748)

This re-adds Secret to the bot token, without losing the `Arc<str>` opt.
This works by using a newtype to Zeroise on last drop, using
`Arc::get_mut`.

GnomedDev committed:
Take Arc<Http> in ClientBuilder::new_with_http (#2749)

This allows the user to make a http, pass it off to a background task,
then use it in serenity, instead of having to initialize two.

GnomedDev committed:
Remove duplicate content_safe implementation (#2759)

GnomedDev committed:
Simplify the message cache (#2757)

GnomedDev committed:
Upgrade to small-fixed-array 0.4 (#2755)

GnomedDev committed:
Avoid HashMap in ShardQueue (#2750)

vicky5124 committed:
Switch to a generic type argument for `ClientBuilder::data` (#2765)

GnomedDev committed:
Fix `VoiceServerUpdate::guild_id` being wrapped in `Option` (#2766)

Noticed by @vicky5124, this does not need to be optional.

https://discord.com/developers/docs/topics/gateway-events#voice-server-update

GnomedDev committed:
Improve MessageBuilder to fit with other builders (#2767)

GnomedDev committed:
Fill message cache with ChannelId::messages and ChannelId::pins (#2768)

GnomedDev committed:
Defer deserialisation of Dispatch to handle errors (#2764)

wackbyte committed:
Remove duplicate `external_emojis` accessor (#2770)

Closes #2769

mkrasnitski committed:
Remove `CreateAttachment::id` field (#2771)

GnomedDev committed:
Fix remaining clippy warnings

arqunis committed:
Fix compilation from rebase

mkrasnitski committed:
Minor adjustments

arqunis committed:
Format unformatted code from rebase

arqunis committed:
Remove rebase artifacts in README

arqunis committed:
Synchronise README example from the `current` branch

cheesycod committed:
Apply Strum derives for FullEvent and Event (#2788)

GnomedDev committed:
Remove `Channel::name` (#2787)

`Channel::name` just called `to_channel` (cloning the entire `Channel`!) and
`PrivateChannel::name` or `GuildChannel::name`.

`GuildChannel::name` just returned the field of the same name, for consistency
with `PrivateChannel::name`.

`PrivateChannel::name` returned a non-standard format string that would be very
easy to replicate in user code.

I also tweaked the signature of `OnlineStatus::name` to stop passing an 8 byte
reference to a 1 byte struct, and return `&'static str`.

GnomedDev committed:
Use join_to_string throughout serenity (#2785)

GnomedDev committed:
Use ToArrayString for int -> str (#2786)

cheesycod committed:
Rename `unstable_discord_api` to `unstable` and disable it by default (#2789)

This also disables the `#[non_exhaustive]` attribute on `FullEvent` if
`unstable` is enabled as a feature.

It is already expected that any change to Serenity with `unstable` enabled may
or may not break code using this feature and making this enum exhaustive allows
for use cases such as fine-grained audit logging, where new variants should be
a compiler error.

Co-authored-by: Gnome! <david2005thomas@Gmail.com>

GnomedDev committed:
Remove deprecated Channel::is_nsfw methods (#2792)

GnomedDev committed:
Replace manual Hash/Eq implementation on User/Role (#2793)

GnomedDev committed:
Remove useless method wrappers (#2795)

GnomedDev committed:
Fix unused variable warning from rebase

GnomedDev committed:
Remove deprecated mapped refs and Cache::channels (#2796)

GnomedDev committed:
Use ExtractMap instead of HashMap when possible (#2797)

GnomedDev committed:
Remove examples and inaccurate docs from HTTP (#2805)

GnomedDev committed:
Use NonMax in HTTP function signatures (#2803)

arqunis committed:
Fix infinite loop in event deserialisation

GnomedDev committed:
Make Guild::shard_id follow additive features (#2813)

This PR:
- Removes the `cache` feature locked function that called shard_count
for the user, since that is useless and breaks the "all features should
be additive" guidance
- Removes the useless example that just showed how to call a function
- Removed duplicate documentation on Guild, PartialGuild, and
InviteGuild just to direct to the canonical documentation where the
function is implemented.

GnomedDev committed:
Remove deprecated `Member` and `Guild` methods (#2817)

Followup to #2816

---------

Co-authored-by: Alex M. M. <acdenissk69@gmail.com>

GnomedDev committed:
Replace `CacheHttp` with `&Http` in more methods (#2818)

This is possible now more methods are cache-less.

GnomedDev committed:
Remove `serde_json::Value` from `Http` (#2806)

This avoids allocating instances of `serde_json::Value`s.

GnomedDev committed:
Add audit log reasons to remaining model methods (#2821)

GnomedDev committed:
Remove even more impl CacheHttp (#2822)

These were found by auditing .http() calls, instead of just manually
looking at `impl CacheHttp`.

UserIsntAvailable committed:
Remove `fs` feature for `tokio` by default (#2824)

I currently working with wasm on one of my personal projects. I'm would
like to use `serenity` to get access to the discord API response models,
but currently the library can not the compiled to wasm, because the
`tokio/fs` feature is not supported in that target.

I went ahead and removed the need to include that feature by default. By
the looks of it, it only was used with the `builder` and `model`
features, so I conditionally added `tokio/fs` back when those features
are being used (the `model` feature already depends on `builder` so that
doesn't need to be changed).

GnomedDev committed:
Update `reqwest` to v0.12 (#2826)

This has to be on the next branch as reqwest is exposed in a couple
places.

GnomedDev committed:
Only include `cache::Settings::time_to_live` if temp cache is enabled (#2833)

GnomedDev committed:
Add resume_gateway_url support (#2832)

This announcement was missed (around August 2022) and probably the cause
of a lot of the disconnections.

jamesbt365 committed:
Correct documentation for `UserId::to_user` (#2809)

The documentation was never updated after #2662 removed the user cache.

NotNorom committed:
Use CreateAttachment for EditGuild::banner (#2844)

GnomedDev committed:
Fix AuditLogEntry::user_id (#2846)

This field is documented as nullable but that isn't reflected in the
model. I also took the opportunity to replace the Option<Vec> with a
default Vec.

mkrasnitski committed:
Restore match arm for `Error::Gateway` variant (#2847)

This was mistakenly removed in #2278 and wasn't caught because of the
wildcard pattern in the match.

GnomedDev committed:
Remove multiple event handler support (#2849)

TheCataliasTNT2k committed:
Use RwLock for active collectors collection (#2851)

jamesbt365 committed:
Provide old presence data from the cache on presence update (#2852)

TheCataliasTNT2k committed:
Allow filtering events before they are dispatched (#2853)

jamesbt365 committed:
Remove permission checks (#2855)

Users should realistically be checking the permissions themselves, or
handling the HTTP error from Discord.

This removes any cases where permission checking inside the library is
broken because of Discord's changes or due to oversights.

This also changes the documentation on the prune functionality, this was
recently changed to also require `MANAGE_GUILDS` as well as
`KICK_MEMBERS`.

GnomedDev committed:
Implement ToArrayString for Mention (#2857)

cheesycod committed:
Add PartialGuild::greater_member_hierarchy + member_highest_role (#2856)

GnomedDev committed:
Update bool_to_bitflags to fix compilation without unstable

GnomedDev committed:
Remove recently deprecated items

jamesbt365 committed:
Remove RUSTFLAGS from .cargo/config.toml (#2875)

This isn't needed anymore because simd-json has been removed on `next`.

mlesin commented:
Are there any plans to merge this?

GnomedDev commented:
As the author seems to be MIA, no. If anyone wants to rebase and solve outstanding reviews, feel free to open a new PR.

qthree commented:
fyi, I cherry-picked commit from this PR on top of `next` branch and added socks4/5 proxy support, feel free to use it https://github.com/qthree/serenity/tree/ws-proxy

Morb0 commented:
a critically important feature
